File: packages/server/database-id.ts
export type DatabaseId = string;

export function requireDatabaseId(value: unknown, label = "databaseId"): DatabaseId {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${label} is required`);
  }

  return value;
}

export function databaseIdFromUrl(url: string | URL): DatabaseId {
  const parsed = typeof url === "string" ? new URL(url, "http://pyre.local") : url;
  return requireDatabaseId(parsed.searchParams.get("databaseId"));
}

export function withDatabaseId<T extends Record<string, unknown>>(databaseId: DatabaseId, message: T): T & { databaseId: DatabaseId } {
  return {
    ...message,
    databaseId: requireDatabaseId(databaseId),
  };
}

File: src/server/database_id.rs
use crate::sync::SyncPageResult;

pub type DatabaseId = String;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DatabaseIdError {
    label: String,
}

impl DatabaseIdError {
    fn new(label: &str) -> Self {
        Self {
            label: label.to_string(),
        }
    }
}

impl std::fmt::Display for DatabaseIdError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} is required", self.label)
    }
}

impl std::error::Error for DatabaseIdError {}

pub fn require_database_id(value: impl AsRef<str>) -> Result<DatabaseId, DatabaseIdError> {
    require_database_id_with_label(value, "databaseId")
}

pub fn require_database_id_with_label(
    value: impl AsRef<str>,
    label: &str,
) -> Result<DatabaseId, DatabaseIdError> {
    let value = value.as_ref();
    if value.trim().is_empty() {
        return Err(DatabaseIdError::new(label));
    }

    Ok(value.to_string())
}

pub fn with_database_id(
    database_id: impl AsRef<str>,
    mut result: SyncPageResult,
) -> Result<SyncPageResult, DatabaseIdError> {
    result.database_id = Some(require_database_id(database_id)?);
    Ok(result)
}
